x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/assignment/TrainedModelAssignmentRebalancer.java

```diff
@@ -138,7 +138,9 @@ static void copyAssignments(AssignmentPlan source, AssignmentPlan.Builder dest,
             Map<AssignmentPlan.Node, Integer> sourceNodeAssignments = source.assignments(deployment).orElse(Map.of());
             for (Map.Entry<AssignmentPlan.Node, Integer> sourceAssignment : sourceNodeAssignments.entrySet()) {
                 AssignmentPlan.Node node = originalNodeById.get(sourceAssignment.getKey().id());
-                dest.assignModelToNode(deployment, node, sourceAssignment.getValue());
+                if (dest.canAssign(deployment, node, sourceAssignment.getValue())) {
+                    dest.assignModelToNode(deployment, node, sourceAssignment.getValue());
```

Comment:
Is there a potential race condition here, where at the time we check if we can assign the model there is sufficient memory, but when we go to actually assign it, memory use has changed and we still encounter the `IllegalArgumentException`?

Since all calls to `assignModelToNode()` are now wrapped in a `canAssign()` check, does it make sense to move that check into the `assignModelToNode()` method and remove the memory check that throws the `IllegalArgumentException`, since it's no longer reachable?

Reply:
Race condition is not possible, since the rebalancing and assignment planning happen in a single thread. 

`assignModelToNode()` is used in multiple other locations, besides `TraindModelAssignmentRebalancer` and `ZoneAwareAssignmentPlanner`. Therefore, I cannot combine `canAssign()` and `assignModelToNode()` together.

Reply:
I moved `canAssign()` into `assignModelToNode()` in c72ba42